File: python/lib/problem41.py
'''
Problem 41 - Pandigital Prime

We shall say that an n-digit number is pandigital if it makes use of all the digits 1 to n exactly
once. For example, 2143 is a 4-digit pandigital and is also prime.

What is the largest n-digit pandigital prime that exists?
'''

from lib.helpers.runtime import print_answer_and_elapsed_time
from lib.helpers.numbers import sieve_of_eratosthenes

def largest_pandigital_prime():
    maximum = 7654321 # 8- and 9-digit pandigital primes are divisible by 3
    primes = sieve_of_eratosthenes(maximum)
    for number in sorted(primes, reverse=True):
        number_string = ''.join(sorted(str(number)))
        number_length = len(number_string)
        if ''.join(sorted(str(number))) == '1234567'[:number_length]:
            return number

answer = largest_pandigital_prime

if __name__ == '__main__':
    print_answer_and_elapsed_time(answer)

File: haskell/src/Problem41.hs
{- |
Problem 41 - Pandigital Prime

We shall say that an n-digit number is pandigital if it makes use of all the
digits 1 to n exactly once. For example, 2143 is a 4-digit pandigital and is
also prime.

What is the largest n-digit pandigital prime that exists?
-}

module Problem41
( answer
) where

import Data.Digits (digits)
import Data.List (find, sort)
import Data.Maybe (fromJust)
import Helpers.Numbers (primes)
import Helpers.Runtime (printAnswerAndElapsedTime)

isPandigital :: Int -> Bool
isPandigital number = sort numberDigits == take (length numberDigits) [1..]
    where numberDigits = digits 10 number

answer :: Int
answer = fromJust $ find isPandigital $ reverse $ takeWhile (<= 7654321) primes

main = printAnswerAndElapsedTime answer
